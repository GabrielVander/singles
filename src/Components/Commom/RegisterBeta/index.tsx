import React, {ReactElement, useState} from 'react';
import {Box, Button, Text, TextInput} from 'grommet';
import {useTranslation} from 'react-i18next';
import {useAnalytics, useFirestore} from 'reactfire';
import firebase from 'firebase/app';
import {toast} from 'react-toastify';

const RegisterBeta = (): ReactElement => {
    const {t} = useTranslation(['registerBeta']);
    const firestore = useFirestore();
    const analytics = useAnalytics();

    const [email, setEmail] = useState<string>('');
    const [savingEmail, setSavingEmail] = useState<boolean>(false);
    const [hasFocus, setHasFocus] = useState<boolean>(false);

    async function applyForBeta(): Promise<void> {
        setSavingEmail(() => true);

        if (!email || email.length === 0) {
            setSavingEmail(() => false);
            return;
        }

        if (!emailIsValid()) {
            setSavingEmail(() => false);
            toast.error(t('registerBeta:invalidEmail'));
            return;
        }

        const collection = firestore.collection('betaApplications');

        if (await emailExists(collection)) {
            setSavingEmail(() => false);
            toast.info(t('registerBeta:emailExists'));
            return;
        }

        collection
            .add({
                email,
                addedAt: firebase.firestore.FieldValue.serverTimestamp(),
            })
            .then((value) => {
                analytics.logEvent('appliedForBeta', {
                    email,
                    id: value.id,
                });
                setSavingEmail(() => false);
                toast.success(t('registerBeta:emailSaved'));
            })
            .catch((reason) => {
                setSavingEmail(() => false);
                toast.error(t('registerBeta:error', {reason}));
            });
    }

    function emailIsValid(): boolean {
        return new RegExp('\\S+@\\S+\\.\\S+').test(email);
    }

    async function emailExists(collection: firebase.firestore.CollectionReference): Promise<boolean> {
        const querySnapshot = await collection.where('email', '==', email).get();

        return !querySnapshot.empty;
    }

    // noinspection HtmlUnknownTarget
    return (
        <Box direction="row" width="large" justify="center">
            <Box
                direction="row"
                align="center"
                width="medium"
                border={{
                    side: 'bottom',
                    color: hasFocus ? 'focus' : 'brand',
                }}
            >
                <TextInput
                    plain
                    placeholder={<Text size="small">{t('registerBeta:emailPlaceholder')}</Text>}
                    value={email}
                    disabled={savingEmail}
                    onChange={(event): void => setEmail(event.target.value)}
                    onFocus={(): void => setHasFocus(true)}
                    onBlur={(): void => setHasFocus(false)}
                />
            </Box>
            <Button disabled={savingEmail} onClick={applyForBeta}>
                <Box
                    round="xlarge"
                    background="accent-1"
                    pad={{
                        vertical: 'small',
                        horizontal: 'medium',
                    }}
                >
                    <Text size="small" color="brand" weight="bold" textAlign="center">
                        {t('registerBeta:applyButton')}
                    </Text>
                </Box>
            </Button>
        </Box>
    );
};

export default RegisterBeta;
