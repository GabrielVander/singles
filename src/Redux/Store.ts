import {createStore} from "redux";
import rootReducer from "./Reducers";

class Store {
    static INITIAL_GLOBAL_STATE = {};
    // @ts-ignore
    static store = createStore(rootReducer, Store.INITIAL_GLOBAL_STATE, window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__());
}

export default Store;
