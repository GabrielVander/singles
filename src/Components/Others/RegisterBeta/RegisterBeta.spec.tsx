import {render} from "@testing-library/react";
import {Provider} from "react-redux";
import Store from "../../../Redux/Store";
import {BrowserRouter} from "react-router-dom";
import React from "react";
import RegisterBeta from "./index";

beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
        writable: true,
        value: jest.fn().mockImplementation(query => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        }))
    });
});

test('matches snapshot', () => {
    const component = render(
        <Provider store={Store.store}>
            <BrowserRouter>
                <RegisterBeta/>
            </BrowserRouter>
        </Provider>
    );
    expect(component).toMatchSnapshot();
})
