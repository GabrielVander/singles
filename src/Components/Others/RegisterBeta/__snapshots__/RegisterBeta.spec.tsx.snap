// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="StyledBox-sc-13pk1d4-0 gljNXM"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 bXnNfl"
        >
          <div
            class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
          >
            <div
              class="StyledTextInput__StyledPlaceholder-sc-1x30a0s-2 hIRoSf"
            >
              <span
                class="StyledText-sc-1sadyjn-0 faTntD"
              >
                registerBeta:emailPlaceholder
              </span>
            </div>
            <input
              autocomplete="off"
              class="StyledTextInput-sc-1x30a0s-0 ishQAc"
              value=""
            />
          </div>
        </div>
        <button
          class="StyledButton-sc-323bzc-0 kUDdtj"
          type="button"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 Szkkv"
          >
            <span
              class="StyledText-sc-1sadyjn-0 cmOUuj"
            >
              registerBeta:applyButton
            </span>
          </div>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="StyledBox-sc-13pk1d4-0 gljNXM"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 bXnNfl"
      >
        <div
          class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 flNVco"
        >
          <div
            class="StyledTextInput__StyledPlaceholder-sc-1x30a0s-2 hIRoSf"
          >
            <span
              class="StyledText-sc-1sadyjn-0 faTntD"
            >
              registerBeta:emailPlaceholder
            </span>
          </div>
          <input
            autocomplete="off"
            class="StyledTextInput-sc-1x30a0s-0 ishQAc"
            value=""
          />
        </div>
      </div>
      <button
        class="StyledButton-sc-323bzc-0 kUDdtj"
        type="button"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 Szkkv"
        >
          <span
            class="StyledText-sc-1sadyjn-0 cmOUuj"
          >
            registerBeta:applyButton
          </span>
        </div>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
